File: 497_RandomPointsinRect.java
import java.util.Arrays;

class RandomPointsRect {
  public static void main(String[] args) {
    int[][] rects = new int[][] {{1,1,5,5}, {10,10, 13, 13}};
    RandomPointsRect obj = new RandomPointsRect(rects);
    System.out.println(obj.cumsum[2]);
    for (int i=0; i<5; ++i) {
      int[] output = obj.pick();
      System.out.println(output[0]+","+output[1]);
    }
  }
  int[][] rects;
  int[] cumsum;
  public RandomPointsRect(int[][] rects) {
    cumsum = new int[rects.length+1];

    for (int i=1; i<=rects.length; ++i) {
      cumsum[i]=cumsum[i-1] + size(rects[i-1]);
    }

    this.rects = rects;
  }

  private int size(int[] rect) {
    return (rect[3]-rect[1]+1) * (rect[2]-rect[0]+1);
  }

  private int[] draw(int[] rect, int idx) {
    return new int[] {rect[0]+idx%(rect[2]-rect[0]+1), rect[1]+idx/(rect[2]-rect[0]+1)};
  }
  
  public int[] pick() {
    int rand = (int) (Math.random() * cumsum[rects.length]);
    int idx = Arrays.binarySearch(cumsum, rand);
    if (idx<0) idx=-idx-2;
    return draw(rects[idx], rand-cumsum[idx]);
  }
}

File: 497_RandomPointsinRect.cpp
#include <vector>
#include <iostream>
#include <iterator>
#include <stdlib.h>

using namespace std;

class Solution {
  private:
    int n;
    vector<int> acc;
    vector<vector<int>> rects;
    int binary_search(int num) {
      int l=0, r=n, m;
      while (l<r) {
        m = l + (r-l)/2 + (r-l)%2;
        if (acc[m]>num)
          r = m-1;
        else
          l = m;
      }
      return l;
    }
  public:
    Solution(vector<vector<int>>& rects) {
      n = rects.size();
      this->rects = rects;
      acc.push_back(0);
      for (auto &rect : rects) 
        acc.push_back(acc.back() + (rect[2]-rect[0]+1) * (rect[3]-rect[1]+1));
    }
    
    vector<int> pick() {
      int num = rand() % acc.back();
      int idx = binary_search(num);
      num -= acc[idx];
      auto &rect = rects[idx];

      int l = rect[2] - rect[0] + 1;
      return {rect[0] + num % l, rect[1] + num / l};
    }
};

int main() {
  vector<vector<int>> rects = {{-2,-2,-1,-1}, {1,0,3,0}};
  Solution sol(rects);
  vector<int> res;
  for (int i=0; i<3; ++i) {
    res = sol.pick();
    copy(res.begin(), res.end(), ostream_iterator<int>(cout, ", "));
  }
  return 0;
}